qt/widgets/common/src/FitScriptGeneratorDataTable.cpp

```diff
@@ -0,0 +1,308 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2020 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidQtWidgets/Common/FitScriptGeneratorDataTable.h"
+
+#include <QAbstractItemView>
+#include <QDoubleValidator>
+#include <QHeaderView>
+#include <QHoverEvent>
+#include <QLineEdit>
+#include <QStringList>
+#include <QValidator>
+
+namespace {
+
+int WS_INDEX_MIN(0);
+int WS_INDEX_MAX(100000);
+double X_EXTENT(100000.0);
+int X_PRECISION(5);
+
+QStringList const COLUMN_HEADINGS({"Name", "WS Index", "StartX", "EndX"});
+QString const TABLE_STYLESHEET("QTableWidget {\n"
+                               "    font-size: 8pt;\n"
+                               "    border: 1px solid #828790;\n"
+                               "}\n"
+                               "\n"
+                               "QTableWidget::item:selected {\n"
+                               "    background-color: #c7e0ff;\n"
+                               "    color: #000000;\n"
+                               "}"
+                               "\n"
+                               "QTableWidget::item:hover {\n"
+                               "    background-color: #c7e0ff;\n"
+                               "    color: #000000;\n"
+                               "}");
+
+QValidator *createXValidator() {
+  auto validator = new QDoubleValidator(-X_EXTENT, X_EXTENT, X_PRECISION);
+  validator->setNotation(QDoubleValidator::Notation::StandardNotation);
+  return validator;
+}
+
+QValidator *createWSIndexValidator() {
+  return new QIntValidator(WS_INDEX_MIN, WS_INDEX_MAX);
+}
+
+QTableWidgetItem *createTableItem(QString const &value,
+                                  Qt::AlignmentFlag const &alignment,
+                                  bool editable) {
+  auto item = new QTableWidgetItem(value);
+  item->setTextAlignment(alignment);
+  if (!editable)
+    item->setFlags(item->flags() ^ Qt::ItemIsEditable);
```

Comment:
what is the ^ for?

Reply:
It combines `item->flags()` and `Qt::ItemIsEditable` into the same `Qt::ItemFlags` object